components/places/src/bookmark_sync/store.rs

```diff
@@ -0,0 +1,1020 @@
+            SELECT guid, 1, 1 FROM moz_bookmarks_deleted",
+        )?;
+
+        Ok(())
+    }
+}
+
+impl<'a> Store for BookmarksStore<'a> {
+    #[inline]
+    fn collection_name(&self) -> &'static str {
+        "bookmarks"
+    }
+
+    fn apply_incoming(
+        &self,
+        inbound: IncomingChangeset,
+        incoming_telemetry: &mut telemetry::EngineIncoming,
+    ) -> result::Result<OutgoingChangeset, failure::Error> {
+        use dogear::Store;
+
+        // Stage all incoming items.
+        let timestamp = inbound.timestamp;
+        let mut tx = self
+            .db
+            .time_chunked_transaction(Duration::from_millis(1000))?;
+        for incoming in inbound.changes {
+            self.apply_payload(timestamp, incoming.0)?;
+            tx.maybe_commit()?;
+        }
+        tx.commit()?;
+
+        // write the timestamp now, so if we are interrupted merging or
+        // creating outgoing changesets we don't need to re-download the same
+        // records.
+        put_meta(self.db, LAST_SYNC_META_KEY, &(timestamp.as_millis() as i64))?;
+
+        // Merge and stage outgoing items.
+        self.merge_with_driver(&Driver)?;
+
+        let mut outgoing = OutgoingChangeset::new(self.collection_name().into(), inbound.timestamp);
+        // TODO: Fetch staged items from `itemsToUpload`.
+        Ok(outgoing)
+    }
+
+    fn sync_finished(
+        &self,
+        new_timestamp: ServerTimestamp,
+        records_synced: &[String],
+    ) -> result::Result<(), failure::Error> {
+        unimplemented!("TODO: Write sync records back to the mirror and update sync statuses");
+    }
+
+    fn get_collection_request(&self) -> result::Result<CollectionRequest, failure::Error> {
+        let since = get_meta::<i64>(self.db, LAST_SYNC_META_KEY)?
+            .map(|millis| ServerTimestamp(millis as f64 / 1000.0))
+            .unwrap_or_default();
+        Ok(CollectionRequest::new(self.collection_name())
+            .full()
+            .newer_than(since))
+    }
+
+    fn reset(&self) -> result::Result<(), failure::Error> {
+        unimplemented!("TODO: Wipe staged items and reset sync statuses");
+    }
+
+    fn wipe(&self) -> result::Result<(), failure::Error> {
+        log::warn!("not implemented");
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::api::places_api::{test::new_mem_api, ConnectionType};
+    use crate::bookmark_sync::store::BookmarksStore;
+    use crate::tests::{
+        assert_json_tree as assert_local_json_tree, insert_json_tree as insert_local_json_tree,
+    };
+    use dogear::{Store, Validity};
+    use serde_json::json;
+
+    use std::cell::Cell;
+    use sync15::Payload;
+
+    #[test]
+    fn test_fetch_remote_tree() -> Result<()> {
+        let records = vec![
+            json!({
+                "id": "qqVTRWhLBOu3",
+                "type": "bookmark",
+                "parentid": BookmarkRootGuid::Unfiled.as_guid(),
```

Comment:
I _think_ this might be wrong. `parentid` (and `id`) should use [this format](https://searchfox.org/mozilla-central/rev/23fb3d3791eb9155fb31e76d322d6812cffded90/toolkit/components/places/PlacesSyncUtils.jsm#65-72) for roots, not the Places GUID format with underscores.

Reply:
Nice catch - I updated the "incoming" tests to use the correct variant